File: src/lib/ipc/subscriber.ts
import { invoke } from '@tauri-apps/api/core';

export interface MqttMonitorPublisherDto {
  id: string;
  broker: string;
  port: number;
  topic: string;
}

export interface MqttMonitorStatusDto {
  connected: boolean;
  subscribing: boolean;
  include_sys: boolean;
  publisher_id: string | null;
  broker: string;
  port: number;
  topic_filter: string;
  message_count: number;
  last_message_at: string | null;
  last_error: string | null;
}

export interface MqttMonitorMessageDto {
  timestamp: string;
  topic: string;
  payload: string;
  qos: 0 | 1 | 2;
  retain: boolean;
}

export interface StartMqttMonitorRequest {
  publisher_id: string;
  topic_filter: string;
  include_sys: boolean;
}

type MqttMonitorStatusDtoRaw = {
  connected: boolean;
  subscribing: boolean;
  includeSys: boolean;
  publisherId?: string | null;
  broker: string;
  port: number;
  topicFilter: string;
  messageCount: number;
  lastMessageAt?: string | null;
  lastError?: string | null;
};

function normalizeStatus(raw: MqttMonitorStatusDtoRaw): MqttMonitorStatusDto {
  return {
    connected: raw.connected,
    subscribing: raw.subscribing,
    include_sys: raw.includeSys,
    publisher_id: raw.publisherId ?? null,
    broker: raw.broker,
    port: raw.port,
    topic_filter: raw.topicFilter,
    message_count: raw.messageCount,
    last_message_at: raw.lastMessageAt ?? null,
    last_error: raw.lastError ?? null,
  };
}

export async function listMqttMonitorPublishers(): Promise<MqttMonitorPublisherDto[]> {
  return invoke('list_mqtt_monitor_publishers');
}

export async function openMqttMonitorWindow(): Promise<void> {
  return invoke('open_mqtt_monitor_window');
}

export async function getMqttMonitorStatus(): Promise<MqttMonitorStatusDto> {
  const raw = await invoke<MqttMonitorStatusDtoRaw>('get_mqtt_monitor_status');
  return normalizeStatus(raw);
}

export async function listMqttMonitorMessages(): Promise<MqttMonitorMessageDto[]> {
  return invoke('list_mqtt_monitor_messages');
}

export async function clearMqttMonitorMessages(): Promise<void> {
  return invoke('clear_mqtt_monitor_messages');
}

export async function startMqttMonitor(
  req: StartMqttMonitorRequest,
): Promise<MqttMonitorStatusDto> {
  const raw = await invoke<MqttMonitorStatusDtoRaw>('start_mqtt_monitor', {
    req: {
      publisherId: req.publisher_id,
      topicFilter: req.topic_filter,
      includeSys: req.include_sys,
    },
  });
  return normalizeStatus(raw);
}

export async function stopMqttMonitor(): Promise<MqttMonitorStatusDto> {
  const raw = await invoke<MqttMonitorStatusDtoRaw>('stop_mqtt_monitor');
  return normalizeStatus(raw);
}

File: src-tauri/src/commands/dto/subscriber.rs
use crate::app_state::{MqttMonitorMessageState, MqttMonitorStatusState};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MqttMonitorPublisherDto {
    pub id: String,
    pub broker: String,
    pub port: u16,
    pub topic: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MqttMonitorStatusDto {
    pub connected: bool,
    pub subscribing: bool,
    pub include_sys: bool,
    pub publisher_id: Option<String>,
    pub broker: String,
    pub port: u16,
    pub topic_filter: String,
    pub message_count: usize,
    pub last_message_at: Option<String>,
    pub last_error: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MqttMonitorMessageDto {
    pub timestamp: String,
    pub topic: String,
    pub payload: String,
    pub qos: u8,
    pub retain: bool,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MqttMonitorTopicNodeDto {
    pub id: String,
    pub label: String,
    pub full_path: String,
    pub has_children: bool,
    pub latest_message: Option<MqttMonitorMessageDto>,
    pub children: Vec<MqttMonitorTopicNodeDto>,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StartMqttMonitorRequest {
    pub publisher_id: String,
    pub topic_filter: String,
    pub include_sys: bool,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GetMqttMonitorTreeRequest {
    pub expanded_paths: Vec<String>,
    pub include_all: bool,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GetMqttMonitorTopicDetailRequest {
    pub full_path: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MqttMonitorTopicDetailDto {
    pub full_path: String,
    pub latest_message: Option<MqttMonitorMessageDto>,
}

impl From<MqttMonitorStatusState> for MqttMonitorStatusDto {
    fn from(value: MqttMonitorStatusState) -> Self {
        Self {
            connected: value.connected,
            subscribing: value.subscribing,
            include_sys: value.include_sys,
            publisher_id: value.publisher_id,
            broker: value.broker,
            port: value.port,
            topic_filter: value.topic_filter,
            message_count: value.message_count,
            last_message_at: value.last_message_at,
            last_error: value.last_error,
        }
    }
}

impl From<MqttMonitorMessageState> for MqttMonitorMessageDto {
    fn from(value: MqttMonitorMessageState) -> Self {
        Self {
            timestamp: value.timestamp,
            topic: value.topic,
            payload: value.payload,
            qos: value.qos,
            retain: value.retain,
        }
    }
}
